designs/statefulset-disruption.md

```diff
@@ -0,0 +1,891 @@
+
+Persistent Storage in Kubernetes involves many moving parts, most of which may not be relevant for the decision at hand. 
+
+For the purpose of this document, you should know that:
+- The **Container Storage Interface** (CSI) is a standard way for Container Orchestrators to provision persistent volumes from storage providers and expose block and file storage systems to containers.
+- The **AttachDetach Controller** watches for stateful workloads that are waiting on their storage, and ensures that their volumes are attached to the right node. Also watches for attached volumes that are no longer in use, and ensures they are detached. 
+- The **CSI Controller** attaches/detaches volumes to nodes with pods that require Persistent Volumes. (I.e. Calls EC2 AttachVolume). [^5]
+- The **CSI Node Service** mounts[^3] volumes to make them available for use by pods. Unmounts volumes after pods terminate to ensure they are no longer in use. Runs on each node. The Kubelet's Volume Manager watches for stateful pods and calls csi node service.
+- The CSI Specification states that the container orchestrator must interact with CSI  Plugin through the following flow of Remote Procedure Calls when a workload requires persistent storage: ControllerPublishVolume (i.e. attach volume to node) -> NodeStageVolume (Mount volume to global node mount-point) -> NodePublishVolume (Mount volume to pod's mountpoint) (and when volume no longer in use: NodeUnbpublishVolume -> NodeUnstageVolume -> ControllerUnpublishVolume) 
+
+For the purpose of this document, assume volumes have already been created and will never be deleted.
+
+<details open>
+<summary>If you want to dive one level deeper, open the dropdown to see the following diagram of what happens between pod eviction and volume detachment</summary>
+<img src="https://github.com/AndrewSirenko/karpenter-provider-aws/assets/68304519/fc577f2c-2de0-4aee-8daa-110a1eb0b990" alt="Stateful Pod Termination" class="inline"/>
+</details>
+
+### Ideal Disruption Flow for Stateful Workloads
+
+In order for a stateful pods to smoothly migrate from the terminating node to another node, the following steps must occur in order:
+
+0. Node marked for deletion
+1. Stateful pods must enter `terminated` state
+2. Volumes must be confirmed as unmounted (By CSI Node Service)
+3. Volumes must be confirmed as detached from instance (By AttachDetach & CSI Controllers)
+4. Karpenter terminates EC2 Instance
+5. Karpenter deletes finalizer on Node
+
+See the following diagram for a more detailed sequence of events.  
+
+![ideal](https://github.com/AndrewSirenko/karpenter-provider-aws/assets/68304519/ae5267e1-fe2d-455f-9199-60cee772f0ae)
+
+### Problems
+
+Today, with customers with default Karpenter `v0.37.0` and EBS CSI Driver `v1.31.0` configurations may experience two different kinds of delays once their disrupted stateful workloads are scheduled on a new node. 
+
+**Problem A. If step 2 doesn't happen, there will be a 6+ minute delay.**
```

Comment:
I'm sure there's a bunch of details here, what was the reason why we have to do 6m here? Does 6+ mean at least 6 minutes of waiting? what's the worst case, infinite?

Reply:
6m was an arbitrary time decided by Kubernetes. Atleast 6 minutes since node is unhealthy until AttachDetach Controller reconciler loop can force a `ControllerUnpublishVolume` detach call. Sometimes it is measured as 7 minutes end-to-end.